### PR TITLE
Handle Empty shadowObserver

### PR DESCRIPTION
**Pull request by Gusted**
Pretty annoying in console

### CODE REVIEW

**Comment by alexanderby** [comment]
Hi @Gusted, could you please share an example URL?

File: src/inject/dynamic-theme/watch.ts [hunk 1]
```diff
@@ -179,9 +179,11 @@ export function watchForStyleChanges(update: (styles: ChangedStyles) => void) {
             return;
         }
         const shadowObserver = new MutationObserver(handleMutations);
-        shadowObserver.observe(node.shadowRoot, mutationObserverOptions);
-        shadowObservers.add(shadowObserver);
-        nodesShadowObservers.set(node, shadowObserver);
+        if (shadowObserver.takeRecords().length != 0) { // Don't handle empty Shadowobservers
+            shadowObserver.observe(node.shadowRoot, mutationObserverOptions);
```
Comment (alexanderby) [inline src/inject/dynamic-theme/watch.ts]: As far as I understand, the error happens when shadowRoot is not an instance of Shadow Root.

**Comment by Gusted** [comment]
open the console in this pull reqeust

File: src/inject/dynamic-theme/watch.ts [hunk 2]
```diff
@@ -179,9 +179,11 @@ export function watchForStyleChanges(update: (styles: ChangedStyles) => void) {
             return;
         }
         const shadowObserver = new MutationObserver(handleMutations);
-        shadowObserver.observe(node.shadowRoot, mutationObserverOptions);
-        shadowObservers.add(shadowObserver);
-        nodesShadowObservers.set(node, shadowObserver);
+        if (shadowObserver.takeRecords().length != 0) { // Don't handle empty Shadowobservers
```
Comment (alexanderby) [inline src/inject/dynamic-theme/watch.ts]: It looks like it will always return 0, since the observer is not observing anything.

Comment (Gusted) [reply src/inject/dynamic-theme/watch.ts]: It's observing a empty {}

**Comment by alexanderby** [comment]
So, it happens for custom elements, that have shadow root attached in close mode, so a simple `element.shadowRoot != null` check somewhere should work.

File: src/inject/dynamic-theme/watch.ts [hunk 3]
```diff
@@ -179,9 +179,11 @@ export function watchForStyleChanges(update: (styles: ChangedStyles) => void) {
             return;
```
Comment (alexanderby) [inline src/inject/dynamic-theme/watch.ts]: So, you can basically add a check for a shadow root in this `if` block.

**Comment by Gusted** [comment]
>?

File: src/inject/dynamic-theme/watch.ts [hunk 4]
```diff
@@ -175,7 +175,7 @@ export function watchForStyleChanges(update: (styles: ChangedStyles) => void) {
     }
 
     function subscribeForShadowRootChanges(node: Element) {
-        if (nodesShadowObservers.has(node)) {
+        if (nodesShadowObservers.has(node) || node.shadowRoot != null) {
```
Comment (alexanderby) [inline src/inject/dynamic-theme/watch.ts]: It should exit, when shadow root is null.

**Comment by alexanderby** [comment]
Thanks for noticing, Travis did break, so I switched to GitHub actions.

**Comment by Gusted** [comment]
Yeah I saw Travis was quite broke has to commited twice to trigger Travis

**Pull request merged**